processor-api/src/main/java/com/synopsys/integration/alert/processor/api/NotificationProcessor.java

```diff
@@ -0,0 +1,121 @@
+        NotificationFilter filter,
+        ProviderMessageExtractor providerMessageExtractor,
+        ProjectMessageDigester projectMessageDigester,
+        ProviderMessageDetailer providerMessageDetailer,
+        ProjectMessageSummarizer projectMessageSummarizer,
+        ProviderMessageDistributor providerMessageDistributor
+    ) {
+        this.filterableNotificationExtractor = filterableNotificationExtractor;
+        this.filter = filter;
+        this.providerMessageExtractor = providerMessageExtractor;
+        this.projectMessageDigester = projectMessageDigester;
+        this.providerMessageDetailer = providerMessageDetailer;
+        this.projectMessageSummarizer = projectMessageSummarizer;
+        this.providerMessageDistributor = providerMessageDistributor;
+    }
+
+    public final void processNotifications(List<AlertNotificationModel> notifications) {
+        processNotifications(notifications, null);
+    }
+
+    public final void processNotifications(List<AlertNotificationModel> notifications, @Nullable FrequencyType frequency) {
+        List<? extends FilterableNotificationWrapper<?>> filterableNotifications = notifications
+                                                                                       .stream()
+                                                                                       .map(filterableNotificationExtractor::wrapNotification)
+                                                                                       .collect(Collectors.toList());
+        Map<DistributionJobModel, List<FilterableNotificationWrapper<?>>> jobsToNotifications = mapJobsToNotifications(filterableNotifications, frequency);
+        for (Map.Entry<DistributionJobModel, List<FilterableNotificationWrapper<?>>> jobToNotifications : jobsToNotifications.entrySet()) {
+            processJobNotifications(jobToNotifications.getKey(), jobToNotifications.getValue());
+        }
+    }
+
+    private void processJobNotifications(DistributionJobModel job, List<FilterableNotificationWrapper<?>> jobNotifications) {
+        ProviderMessageHolder extractedProviderMessages = filter.filter(job, jobNotifications)
+                                                              .stream()
+                                                              .map(providerMessageExtractor::extract)
+                                                              .reduce(ProviderMessageHolder::reduce)
+                                                              .orElse(ProviderMessageHolder.empty());
+
+        ProviderMessageHolder processedProviderMessages = processExtractedNotifications(job, extractedProviderMessages);
+        providerMessageDistributor.distribute(job, processedProviderMessages);
+    }
+
+    protected abstract Map<DistributionJobModel, List<FilterableNotificationWrapper<?>>> mapJobsToNotifications(List<? extends FilterableNotificationWrapper<?>> filterableNotifications, @Nullable FrequencyType frequency);
```

Comment:
Since NotificationProcessor appears to be per notification type, can we have a generic passed to this class and that type would be the type of FilterableNotificationWrapper or something similar?

Reply:
It isn't per notification type. I think it is called with a batch of arbitrary notifications over a specific date range.